File: src/core/spin.ts
// Spin detection. T uses the guideline 3-corner rule with front-corner
// full/mini distinction (the final TST kick, index 4, upgrades mini→full).
// Every other piece uses the tetr.io "all-spin" immobility rule: a piece
// rotated into a spot where it cannot move up, down, left or right is a spin
// and keeps back-to-back.

import { cellsAt, type PieceType, type Rot } from './pieces';
import type { Board } from './board';

export type SpinKind = 'none' | 'mini' | 'full';

// For each T rotation state, the two "front" corner offsets (the side the T
// points toward) and the two "back" corners, relative to the T center.
const FRONT: Record<Rot, [number, number][]> = {
  0: [[-1, 1], [1, 1]],
  1: [[1, 1], [1, -1]],
  2: [[-1, -1], [1, -1]],
  3: [[-1, 1], [-1, -1]],
};
const BACK: Record<Rot, [number, number][]> = {
  0: [[-1, -1], [1, -1]],
  1: [[-1, 1], [-1, -1]],
  2: [[-1, 1], [1, 1]],
  3: [[1, 1], [1, -1]],
};

export function detectSpin(
  board: Board,
  type: PieceType,
  rot: Rot,
  x: number,
  y: number,
  lastMoveWasRotation: boolean,
  lastKickIndex: number,
): SpinKind {
  if (!lastMoveWasRotation) return 'none';
  if (type !== 'T') {
    // all-spin: any non-T piece boxed in on all four sides after a rotation
    return isImmobile(board, type, rot, x, y) ? 'full' : 'none';
  }
  const frontFilled = FRONT[rot].filter(([dx, dy]) => board.filled(x + dx, y + dy)).length;
  const backFilled = BACK[rot].filter(([dx, dy]) => board.filled(x + dx, y + dy)).length;
  if (frontFilled + backFilled < 3) return 'none';
  if (frontFilled === 2) return 'full';
  return lastKickIndex === 4 ? 'full' : 'mini';
}

/** A piece that cannot shift left, right, up or down is "immobile". */
function isImmobile(board: Board, type: PieceType, rot: Rot, x: number, y: number): boolean {
  return board.collides(cellsAt(type, rot, x - 1, y))
    && board.collides(cellsAt(type, rot, x + 1, y))
    && board.collides(cellsAt(type, rot, x, y - 1))
    && board.collides(cellsAt(type, rot, x, y + 1));
}

File: rust/lst-solver/src/spin.rs
//! Spin detection. Port of src/core/spin.ts.

use crate::board::Board;
use crate::pieces::{cells_at, PieceType};

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum SpinKind {
    None,
    Mini,
    Full,
}

impl SpinKind {
    pub fn rank(self) -> u8 {
        match self {
            SpinKind::None => 0,
            SpinKind::Mini => 1,
            SpinKind::Full => 2,
        }
    }
    pub fn as_str(self) -> &'static str {
        match self {
            SpinKind::None => "none",
            SpinKind::Mini => "mini",
            SpinKind::Full => "full",
        }
    }
}

// front[rot] / back[rot]: corner offsets relative to the T center.
const FRONT: [[(i32, i32); 2]; 4] = [
    [(-1, 1), (1, 1)],
    [(1, 1), (1, -1)],
    [(-1, -1), (1, -1)],
    [(-1, 1), (-1, -1)],
];
const BACK: [[(i32, i32); 2]; 4] = [
    [(-1, -1), (1, -1)],
    [(-1, 1), (-1, -1)],
    [(-1, 1), (1, 1)],
    [(1, 1), (1, -1)],
];

pub fn detect_spin(
    board: &Board,
    t: PieceType,
    rot: usize,
    x: i32,
    y: i32,
    last_move_was_rotation: bool,
    last_kick_index: i32,
) -> SpinKind {
    if !last_move_was_rotation {
        return SpinKind::None;
    }
    if t != PieceType::T {
        // all-spin: any non-T boxed in on all four sides after a rotation
        return if is_immobile(board, t, rot, x, y) {
            SpinKind::Full
        } else {
            SpinKind::None
        };
    }
    let front_filled = FRONT[rot]
        .iter()
        .filter(|&&(dx, dy)| board.filled(x + dx, y + dy))
        .count();
    let back_filled = BACK[rot]
        .iter()
        .filter(|&&(dx, dy)| board.filled(x + dx, y + dy))
        .count();
    if front_filled + back_filled < 3 {
        return SpinKind::None;
    }
    if front_filled == 2 {
        return SpinKind::Full;
    }
    if last_kick_index == 4 {
        SpinKind::Full
    } else {
        SpinKind::Mini
    }
}

fn is_immobile(board: &Board, t: PieceType, rot: usize, x: i32, y: i32) -> bool {
    board.collides(&cells_at(t, rot, x - 1, y))
        && board.collides(&cells_at(t, rot, x + 1, y))
        && board.collides(&cells_at(t, rot, x, y - 1))
        && board.collides(&cells_at(t, rot, x, y + 1))
}
